Storybook and TS migration of `PriceSlider` component (#5253)

sunyatasattva opened the pull request:
This PR adds stories for the `PriceSlider` component and migrates it to TypeScript. It also migrates a bunch of associated functions and components.

Fixes #5253

### Testing

### Manual Testing

How to test the changes in this Pull Request:

1. Run `npm run storybook`
2. Explore the [PriceSlider stories](http://localhost:6006/?path=/story/woocommerce-blocks-base-components-priceslider--default). You should see one story: it should be fully interactive, showing all the controls with the appropriate types and defaults.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [x] Same as above

sunyatasattva committed:
Add full documentation to the component props in Storybook

sunyatasattva committed:
Fix `FormattedMonetaryAmount` TypeScript errors

Since this component is passing on props to `NumberFormat`,
it needs to extend all of its original props, except for `onValueChange`
which we wrap in order to accept only `number`s.

sunyatasattva committed:
Convert `constrainRangeSliderValues` to TypeScript

sunyatasattva committed:
Add docs to `Currency` types

sunyatasattva committed:
Convert `PriceSlider` to TypeScript

*Note:* All TypeScript errors were fixed during the conversion,
except for the IE workaround which wasn't clear to me at this time.
Opened an issue: #5276

sunyatasattva committed:
Fix Price Slider story (#5253)

dinhtungdu reviewed (commented):
@sunyatasattva The code looks good. The behavior of the TS version of the block on the frontend is the same as the JS version.

When testing the storybook I noticed some mismatch behavior compared to the block rendered on the frontend:
- I can't input min and max prices directly. But I can delete them.
- If I delete the min price, then delete the max price, the max is set to `0.00` and the min price now becomes negative (`-$2.50`) in my test.

sunyatasattva committed:
Migrate stories for `ProductName` to latest Storybook

Also add props documentation and add named export.

sunyatasattva committed:
Migrate `TotalsFooterItem` to TypeScript and latest Storybook

Also:

 * Add a `LooselyMustHave` utility type.
 * Export `allSettings` so that they can be manipulated in stories and tests

sunyatasattva committed:
Implement a way to easily define and reuse Storybook controls

Implement a currency control for a common use-case of selecting currencies.
It currently implements EUR and USD as they have different properties.

sunyatasattva committed:
Migrate `TotalsDiscount` to TypeScript and implement stories

sunyatasattva committed:
Migrate `TotalsCoupon` to TypeScript and fix stories

sunyatasattva committed:
Remove unused snapshot

sunyatasattva committed:
Change Coupon name within Storybook

sunyatasattva committed:
Nicer handling of removal of a coupon from Storybook

It now dynamically calculates the discount from the actual coupons.

sunyatasattva committed:
Fix TSConfig conflict

sunyatasattva committed:
Merge branch 'trunk' into fix/5311-storybook-cart-checkout

Commit:
Merge commit 'c0c2a215bf2044fb61a77a1c8f629fbdff39eb92' into fix/5253-storybook-priceslider

sunyatasattva committed:
Add stories for no input and with button

sunyatasattva commented:
Hey @dinhtungdu thanks for the review, sorry for taking so long, I was taken by other things and wanted to investigate what you had mentioned.

> * I can't input min and max prices directly. But I can delete them.

You should now be able to input the prices directly. There are no Story controls for the prices, as the component is a [controlled component](https://reactjs.org/docs/forms.html#controlled-components), so that's to be expected in Storybook.

> * If I delete the min price, then delete the max price, the max is set to `0.00` and the min price now becomes negative (`-$2.50`) in my test.

I can reproduce this bug on the frontend. I have reported it: https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5376

dinhtungdu reviewed (approved):
The update looks good to me! :shipit:

sunyatasattva committed:
Merge branch 'trunk' into fix/5253-storybook-priceslider
